### PR TITLE
DEPR: DataFrame.get_dtype_counts

### PR DESCRIPTION
**Pull request by mroeschke**
- [x] xref #18262
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Unfortunately, `.dtype.value_counts()` isn't a perfect replacement for `get_value_counts()` (IMO it's better), so the test replacements calls the internal representation:

```
In [2]: df = pd.DataFrame([[1, 1.5, pd.Timestamp.now(), True]])

In [3]: df
Out[3]:
   0    1                          2     3
0  1  1.5 2019-06-30 11:55:24.380126  True

In [7]: df.get_dtype_counts().index
Out[7]: Index(['float64', 'int64', 'datetime64[ns]', 'bool'], dtype='object')

In [8]: df.dtypes.value_counts().index
Out[8]: Index([datetime64[ns], bool, float64, int64], dtype='object')
```

### CODE REVIEW

File: pandas/tests/frame/test_constructors.py [hunk 1]
```diff
@@ -1693,7 +1693,7 @@ def test_constructor_datetimes_with_nulls(self):
         for arr in [np.array([None, None, None, None,
                               datetime.now(), None]),
                     np.array([None, None, datetime.now(), None])]:
-            result = DataFrame(arr).get_dtype_counts()
+            result = Series(DataFrame(arr)._data.get_dtype_counts())
```
Comment (jreback) [inline pandas/tests/frame/test_constructors.py]: don’t move it to a private method

use .dtypes.value_counts()

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/computation/expressions.py [hunk 2]
```diff
@@ -79,11 +79,11 @@ def _can_use_numexpr(op, op_str, a, b, dtype_check):
             # check for dtype compatibility
             dtypes = set()
             for o in [a, b]:
-                if hasattr(o, 'get_dtype_counts'):
-                    s = o.get_dtype_counts()
+                if hasattr(o, '_data'):
```
Comment (jreback) [inline pandas/core/computation/expressions.py]: can prob just do
o.dtypes.value_counts() (may need a hasattr)

File: pandas/tests/frame/test_api.py [hunk 3]
```diff
@@ -433,9 +433,10 @@ def test_with_datetimelikes(self):
                          'B': timedelta_range('1 day', periods=10)})
         t = df.T
 
-        result = t.get_dtype_counts()
+        #result = Series(t._data.get_dtype_counts())
```
Comment (jreback) [inline pandas/tests/frame/test_api.py]: comment here

File: pandas/core/generic.py [hunk 4]
```diff
@@ -5290,6 +5290,9 @@ def get_dtype_counts(self):
         object     1
         dtype: int64
         """
+        warnings.warn("`get_dtype_counts` has been deprecated and will be "
```
Comment (jreback) [inline pandas/core/generic.py]: can you update the docstring and add deprecated

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by mroeschke** [comment]
All green.

File: pandas/core/generic.py (hunk 4 above)
Comment (TomAugspurger) [reply pandas/core/generic.py]: Can we recommend `.dtypes.value_counts()` here instead? Or... we're in `generic.py` so that may be too hard?

Comment (mroeschke) [reply pandas/core/generic.py]: Yeah unfortunately that solution does not work for `Series`, but I could add `for DataFrames use .dtypes.value_counts()`

File: pandas/core/frame.py [hunk 5]
```diff
@@ -2326,7 +2326,7 @@ def _sizeof_fmt(num, size_qualifier):
             else:
                 _verbose_repr()
 
-        counts = self.get_dtype_counts()
+        counts = self._data.get_dtype_counts()
```
Comment (jreback) [inline pandas/core/frame.py]: this one?

File: pandas/core/generic.py (hunk 4 above)
Comment (jreback) [reply pandas/core/generic.py]: sure, just need something as a replacement (may also want to add in the doc-string itself)

**Review by jreback — CHANGES_REQUESTED** [review]
1 small change & doc-string updates, otherwise lgtm.

File: pandas/core/frame.py (hunk 5 above)
Comment (mroeschke) [reply pandas/core/frame.py]: I think this is okay. It's internal usage and slightly more performant I would think than `dtype.value_counts()` (left as a dictionary as opposed to constructing the Series)

Comment (jreback) [reply pandas/core/frame.py]: can you remove get_dtype_counts() from blocks its unecessary as well

Comment (mroeschke) [reply pandas/core/frame.py]: Looks to be needed to get the dtypes later on for `info`?

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @mroeschke

**Pull request merged**